worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,326 @@
+func (s *subscription) Done() <-chan struct{} {
+	return s.done
+}
+
+// Kill implements worker.Worker.
+func (s *subscription) Kill() {
+	s.tomb.Kill(nil)
+}
+
+// Wait implements worker.Worker.
+func (s *subscription) Wait() error {
+	return s.tomb.Wait()
+}
+
+func (s *subscription) loop() error {
+	for {
+		select {
+		case <-s.tomb.Dying():
+			return tomb.ErrDying
+		}
+	}
+}
+
+type subscriptionOpts struct {
+	*subscription
+	opts []changestream.SubscriptionOption
+}
+
+type eventFilter struct {
+	subscriptionID uint64
+	changeMask     changestream.ChangeType
+	filter         func(changestream.ChangeEvent) bool
+}
+
+// EventQueue defines an event listener and dispatcher for db changes that can
+// be multiplexed to subscriptions. The event queue allows consumers to
+// subscribe via callbacks to the event queue. This is a lockless
+// implementation, all subscriptions and changes are serialized in the main
+// loop. Dispatching is randomized to ensure that subscriptions don't depend on
+// ordering. The subscriptions can be associated with different subscription
+// options, which provide filtering when dispatching. Unsubscribing is provided
+// per subscription, which is done asynchronously.
+type EventQueue struct {
+	catacomb catacomb.Catacomb
+	stream   Stream
+	logger   Logger
+
+	subscriptions      map[uint64]*subscription
+	subscriptionsByNS  map[string][]*eventFilter
+	subscriptionsAll   map[uint64]struct{}
+	subscriptionsCount uint64
+
+	// (un)subscription related channels to serialize adding and removing
+	// subscriptions. This allows the queue to be lock less.
+	subscriptionCh   chan subscriptionOpts
+	unsubscriptionCh chan uint64
+}
+
+// New creates a new EventQueue that will use the Stream for events.
+func New(stream Stream, logger Logger) (*EventQueue, error) {
+	queue := &EventQueue{
+		stream:             stream,
+		logger:             logger,
+		subscriptions:      make(map[uint64]*subscription),
+		subscriptionsByNS:  make(map[string][]*eventFilter),
+		subscriptionsAll:   make(map[uint64]struct{}),
+		subscriptionsCount: 0,
+
+		subscriptionCh:   make(chan subscriptionOpts),
+		unsubscriptionCh: make(chan uint64),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &queue.catacomb,
+		Work: queue.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return queue, nil
+}
+
+// Subscribe creates a new subscription to the event queue. Options can be
+// provided to allow filter during the dispatching phase.
+func (q *EventQueue) Subscribe(handler changestream.Handler, opts ...changestream.SubscriptionOption) (changestream.Subscription, error) {
+	// Get a new subscription count without using any mutexes.
+	subID := atomic.AddUint64(&q.subscriptionsCount, 1)
+
+	sub := newSubscription(subID, handler, func() { q.unsubscribe(subID) })
+	if err := q.catacomb.Add(sub); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	done := make(chan struct{})
+	go func() {
```

Comment:
this go routine makes no sense... just drop the done channel and send to the subCh in the outer select.